qdax/core/emitters/mees_emitter.py

```diff
@@ -0,0 +1,613 @@
+
+
+class MEESEmitterState(EmitterState):
+    """Emitter State for the MAP-Elites-ES emitter."""
+
+    learning_rate: float
+    initial_optimizer_state: optax.OptState  # only used for ADAM
+    optimizer_state: optax.OptState  # only used for ADAM
+    gradient_offspring: Genotype  # Offspring generated through gradient
+    generation_count: int  # Generation counter
+    novelty_archive: Descriptor  # Used to compute novelty for explore
+    last_updated_genotypes: Genotype  # Used to sample parents
+    last_updated_fitnesses: Fitness
+    last_updated_position: jnp.array
+    random_key: RNGKey
+
+
+class MEESEmitter(Emitter):
+    """
+    Emitter reproducing the MAP-Elites-ES exploit-explore and MAP-Elites-ES exploit
+    algorithms from "Scaling MAP-Elites to Deep Neuroevolution" by Colas et al:
+    https://dl.acm.org/doi/pdf/10.1145/3377930.3390217
+
+    One can choose between the two algorithms by setting the use_explore param:
+        ME-ES exploit-explore: alternates between num_optimizer_steps of fitness
+            gradients and num_optimizer_steps of novelty gradients.
+        ME-ES eploit: only uses fitness gradient and no novelty gradients, resample
+            parent from the archive every num_optimizer_steps steps.
+    """
+
+    def __init__(
+        self,
+        config: MEESConfig,
+        total_generations: int,
+        scoring_fn: Callable[
+            [Genotype, RNGKey], Tuple[Fitness, Descriptor, ExtraScores, RNGKey]
+        ],
+        num_descriptors: int,
+    ) -> None:
+        """Initialise the MAP-Elites-ES emitter.
+
+        WARNING: total_generations is required to build the novelty archive.
+
+        Args:
+            config
+            total_generations: total number of generations for which the
+                emitter will run, allow to initialise the novelty archive.
+            scoring_fn: used to evaluate the samples for the gradient estimate.
+            num_descriptors: dimension of the descriptors, used to initialise
+                the empty novelty archive.
+
+        Returns: /
+        """
+        self._config = config
+        self._scoring_fn = scoring_fn
+        self._total_generations = total_generations
+        self._num_descriptors = num_descriptors
+
+        # Initialise optimizer
+        if self._config.adam_optimizer:
+            self._optimizer = optax.adam(learning_rate=config.learning_rate)
+        else:
+            self._optimizer = None
+
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
+    def init(
+        self, init_genotypes: Genotype, random_key: RNGKey
+    ) -> Tuple[MEESEmitterState, RNGKey]:
+        """Initializes the emitter state.
+
+        Args:
+            init_genotypes: The initial population.
+            random_key: A random key.
+
+        Returns:
+            The initial state of the MEESEmitter, a new random key.
+        """
+        assert (
```

Comment:
What does it mean? What is the error? (can you make the error message more explicit?)

Reply:
I agree the assert is not necessary in the init, I replaced it. However, we still need it for the update: ME-ES has batch-size 1 and would not work with a bigger batch-size. I modified the message in the update() function to try to make it more explicit.